### PR TITLE
Add string_concat machinery

### PR DESCRIPTION
**Pull request by jonesmz**

### CODE REVIEW

File: src/osp/string_concat.h [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+ * \returns Nothing.
+ */
+template<typename DESTINATION_T, typename ... STRS_T>
+constexpr void string_append(DESTINATION_T & dest, const STRS_T& ... others)
+{
+    // C++17 fold for summation
+    dest.reserve( ( string_size(dest) + ... + string_size(others) ) );
+
+    // C++17 fold for function calls.
+    ( dest.append(string_data(others), string_size(others)), ... );
```
Comment (z-adams) [inline src/osp/string_concat.h:117]: I have no idea what this does, can I take your word that it works? If so we should merge this so I can use it in my PR

Comment (jonesmz) [reply src/osp/string_concat.h:117]: Basically it expands the parameter pack.

Imagine you have others = string1, string2, string3.

If so, then

( dest.append(string_data(others), string_size(others)), ... );

turns into
dest.append(string_data(string1), string_size(string1));
dest.append(string_data(string2), string_size(string2));
dest.append(string_data(string3), string_size(string3));

Comment (jonesmz) [reply src/osp/string_concat.h:117]: As far as I'm aware, this should work, but obviously if you run into problems with it let me know.

Comment (z-adams) [reply src/osp/string_concat.h:117]: @Capital-Asterisk you can merge this whenever you get a chance to look at it

**Review by z-adams — APPROVED** [review]

**Review by Capital-Asterisk — APPROVED** [review]
This is pretty neat

**Pull request merged**